tools/chpl-language-server/src/symbol_signature.py

```diff
@@ -17,154 +17,262 @@
 # limitations under the License.
 #
 
-from typing import Optional, Union
+from typing import List, Optional, Union
 import chapel
+from dataclasses import dataclass
+import enum
 
 
-# TODO: for now, just text based. but should resolve generic types
-def get_symbol_signature(node: chapel.AstNode) -> str:
-    """
-    get a string representation of a AstNode's signature
+class ComponentTag(enum.Enum):
+    STRING = enum.auto()
+    TYPE = enum.auto()
+    PARAM_VALUE = enum.auto()
+    PLACEHOLDER = enum.auto()
 
-    Note: this is purely textual and does not resolve generic types
-    """
+
+@dataclass
+class Component:
+    tag: ComponentTag
+    value: Union[str, None]
+    node: Optional[chapel.AstNode] = None
+
+    @staticmethod
+    def to_string(comps: List["Component"]) -> str:
+        s = ""
+        for comp in comps:
+            if comp.tag in (
+                ComponentTag.STRING,
+                ComponentTag.TYPE,
+                ComponentTag.PARAM_VALUE,
+            ):
+                if comp.value is not None:
+                    s += str(comp.value)
+            elif comp.tag == ComponentTag.PLACEHOLDER:
+                s += "<...>"
+            else:
+                assert False, f"unknown component tag {comp.tag}"
+        return s
+
+
+def _wrap_str(s: str) -> Component:
+    return Component(ComponentTag.STRING, s)
+
+
+class SymbolSignature:
+    def __init__(self, node: chapel.AstNode):
+        self.node = node
+        self._signature = _get_symbol_signature(self.node)
+
+    def compute_type(self):
+        """
+        if types aren't textually present, try to resolve them
+
+        Note: this is a temporary method that will go away when the resolver is fully online
+        """
+        for i in range(len(self._signature)):
+            tag = self._signature[i].tag
+            node = self._signature[i].node
+            if tag == ComponentTag.TYPE and node is not None:
```

Comment:
Consider a helper method for this: "find piece with tag".

Reply:
There could be multiple Type tags (although not currently), I would prefer to leave it as is